### PR TITLE
[cxx-interop] Fix two headers to be valid in C++ mode.

### PR DESCRIPTION
**Pull request by zoecarver**
This is exported as a C header but in C++ mode `ucal.h` will expose templates that aren't valid. This will hopefully fix the Linux build for https://github.com/apple/swift/pull/32721.

### CODE REVIEW

**Comment by spevans** [comment]
@swift-ci test

**Comment by zoecarver** [comment]
@swift-ci please test.

**Comment by zoecarver** [comment]
I don't have commit access. Could someone review/merge this for me?

**Review by MaxDesiatov — APPROVED** [review]
This seems legit to me, but as there's a way to test the toolchain PR with this without merging, I'll defer to the rest of the requested reviewers to merge it, unless it's urgent.

**Comment by zoecarver** [comment]
> This seems legit to me, but as there's a way to test the toolchain PR with this without merging, I'll defer to the rest of the requested reviewers to merge it, unless it's urgent.

Not urgent. I'm just eager to get that other patch in 😁 I had forgotten that toolchain PR testing works, thanks for running that.

**Comment by zoecarver** [comment]
Friendly ping.

**Comment by MaxDesiatov** [comment]
@compnerd @millenomi would you have a moment to take a look, or would you mind if I merge this otherwise?

**Comment by compnerd** [comment]
Hmm, where is `CFCalendar_Internal.h` used in a C++ context?  I also have the same question for the `CFForSwiftFoundationOnly.h` header.  I think we should make that more explicit in the commit message at the very least.

**Comment by zoecarver** [comment]
> Hmm, where is CFCalendar_Internal.h used in a C++ context? I also have the same question for the CFForSwiftFoundationOnly.h header.

When C++ interop is enabled, any time these are used it's in a C++ context. The same is true for Swift shims (we had to make similar fixes there a while back).

**Comment by zoecarver** [comment]
In other words, it's used in C++ mode any time a Swift program imports `CoreFoundation` and has the flag `-enable-cxx-interop`.

I'll update the commit messages to clarify that this is needed to import these into programs that use the `-enable-cxx-interop` flag.

**Comment by MaxDesiatov** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
I've updated the commit message. @compnerd any other comments?

**Comment by MaxDesiatov** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
@swift-ci please test Linux platform.

**Comment by zoecarver** [comment]
@swift-ci please test Linux platform.

**Comment by zoecarver** [comment]
@swift-ci please test Linux platform.

**Comment by zoecarver** [comment]
@swift-ci please test Linux platform.

**Comment by zoecarver** [comment]
Friendly ping.

**Comment by zoecarver** [comment]
@swift-ci please smoke test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Comment by zoecarver** [comment]
@millenomi friendly ping. Any comments?

**Comment by millenomi** [comment]
I'm a little thick but I'm not sure I get the change for _Internal, which should not be imported outside of the CF build itself, which is in C mode and not C++. Okay on the ForSwift… because that ends up in the CF module, tho.

File: CoreFoundation/Locale.subproj/CFCalendar_Internal.h
Comment (millenomi) [inline CoreFoundation/Locale.subproj/CFCalendar_Internal.h:22]: Per comment: please explain here.

**Review by millenomi — CHANGES_REQUESTED** [review]

File: CoreFoundation/Locale.subproj/CFCalendar_Internal.h
Comment (zoecarver) [reply CoreFoundation/Locale.subproj/CFCalendar_Internal.h:22]: With C++ Interoperability enabled, all headers are imported as C++ headers. So they will all be imported in C++ mode. Even if they're not publically exposed. In other words, all headers in the CF module will be imported as C++ headers in a program that enables C++ interoperability. 

You're right, I think ForSwift may be the only place this is causing issues. To be honest, I know next to nothing about core foundation so, maybe this isn't the correct fix. Feel free to suggest an alternative. 

Anyway, does that answer your question? Would you also like me to put a comment in "CFCalendar_Internal.h" explaining this?

Comment (zoecarver) [reply CoreFoundation/Locale.subproj/CFCalendar_Internal.h:22]: @millenomi friendly ping. I'd really like to get this moving.

Comment (millenomi) [reply CoreFoundation/Locale.subproj/CFCalendar_Internal.h:22]: My bad, I missed your ping back.

**Review by millenomi — APPROVED** [review]

**Comment by millenomi** [comment]
@swift-ci please test and merge

**Comment by millenomi** [comment]
In the interest of getting this going, I'm okay with the safety change on the internal header.

File: CoreFoundation/Locale.subproj/CFCalendar_Internal.h
Comment (zoecarver) [reply CoreFoundation/Locale.subproj/CFCalendar_Internal.h:22]: No worries. I missed your comment the first time ;) Thanks!

**Comment by zoecarver** [comment]
> In the interest of getting this going, I'm okay with the safety change on the internal header.

Great. If you'd like to keep discussing, I'm happy to. (And if we arrive at a better solution, I'm happy to implement it.)

**Pull request merged**
